src/main/java/org/gridsuite/gateway/filters/TokenValidatorGlobalPreFilter.java

```diff
@@ -109,29 +125,71 @@ public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
         ClientID clientID = new ClientID(jwtClaimsSet.getAudience().get(0));
 
         JWSAlgorithm jwsAlg = JWSAlgorithm.parse(jwt.getHeader().getAlgorithm().getName());
-        return gatewayService.getJwksUrl(jwtClaimsSet.getIssuer())
-                .flatMap(jwkSet -> {
-                    try {
-                        URL jwkSetURL = new URL(jwkSet);
-
-                        // Create validator for signed ID tokens
-                        IDTokenValidator validator = new IDTokenValidator(iss, clientID, jwsAlg, jwkSetURL);
-
-                        validator.validate(jwt, null);
-                        // we can safely trust the JWT
-                        LOGGER.debug("Token verified, it can be trusted");
-
-                        //we add the subject header
-                        exchange.getRequest()
-                                .mutate()
-                                .headers(h -> h.set(HEADER_USER_ID, jwtClaimsSet.getSubject()));
-
-                        return chain.filter(exchange);
-                    } catch (JOSEException | BadJOSEException | MalformedURLException e) {
-                        LOGGER.info("{}: 401 Unauthorized, The token cannot be trusted : {}", exchange.getRequest().getPath(), e.getMessage());
-                        return completeWithCode(exchange, HttpStatus.UNAUTHORIZED);
-                    }
+        return jwkUriCache.get(iss.getValue()) != null ? proceedFilter(new FilterInfos(exchange, chain, jwt, jwtClaimsSet, iss, clientID, jwsAlg, jwkUriCache.get(iss.getValue()))) : gatewayService.getJwksUrl(jwtClaimsSet.getIssuer())
```

Comment:
I'd prefere not searching twice, so I'd go with a local variable, even though Jamal may not approve ;-)
And here, as either branch is wordy, I'd go with a 'if' in stead of a ternary.
And may be yet, leave old return as it was, just "intercepting" if the cache contains what we are looking for.

Reply:
ill change the code to use local variable unstead